experimental/hip/pending_queue_actions.c

```diff
@@ -605,27 +638,32 @@ static iree_status_t iree_hal_hip_pending_queue_actions_issue_execution(
   return iree_ok_status();
 }
 
-// Releases resources after completing the given kernel dispatch |action|.
-static void iree_hal_hip_pending_queue_actions_cleanup_execution(
+// Performs the given cleanup |action| on the CPU.
+static iree_status_t iree_hal_hip_pending_queue_actions_issue_cleanup(
     iree_hal_hip_queue_action_t* action) {
   iree_hal_hip_pending_queue_actions_t* actions = action->owning_actions;
   iree_allocator_t host_allocator = actions->host_allocator;
   IREE_TRACE_ZONE_BEGIN(z0);
 
+  // First call user provided callback before releasing any resource.
   if (action->cleanup_callback) {
     action->cleanup_callback(action->callback_user_data);
   }
 
+  // Then release all retained resources.
```

Comment:
nit: ordering comments shouldn't require context as new lines of code added can make things less clear - usually best to state the condition (`// Only release resources after callbacks have been issued.`)